Marksman/CodeActions.fs

```diff
@@ -0,0 +1,83 @@
+module Marksman.CodeActions
+
+open Ionide.LanguageServerProtocol.Types
+open Ionide.LanguageServerProtocol.Logging
+
+open FSharpPlus.GenericBuilders
+open Marksman.Toc
+open Marksman.Workspace
+open Marksman.Text
+open Marksman.Misc
+
+open type System.Console
+
+open type System.Environment
+
+let private logger = LogProvider.getLoggerByName "CodeActions"
+
+type DocumentAction = { name: string; newText: string; edit: Range }
+
+let documentEdit range text documentUri : WorkspaceEdit =
+    let textEdit = { NewText = text; Range = range }
+
+    let mp = Map.ofList [ documentUri, [| textEdit |] ]
+
+    { Changes = Some mp; DocumentChanges = None }
+
+let tableOfContents (document: Doc) : DocumentAction option =
+    monad' {
+        let! toc = Toc.TableOfContents.mk document.index
+
+        let rendered = TableOfContents.render toc
+        let existing = TableOfContents.detect document.text
+
+        let name =
+            match existing with
+            | None -> "Create a Table of Contents"
+            | _ -> "Update the Table of Contents"
+
+        let insertionPoint =
+            match existing with
+            | Some (range) -> Replacing range
+            | None -> TableOfContents.insertionPoint document
+
+        logger.trace (
+            Log.setMessage ("Determining table of contents insertion point")
+            >> Log.addContext "insertionPoint" insertionPoint
+            >> Log.addContext "existing" existing
+            >> Log.addContext "text" rendered
+        )
+
+        let isEmpty lineNumber = document.text.LineContent(lineNumber).Trim().Length.Equals(0)
```

Comment:
nit: `Lenth.Equals(0)` -> `IsEmpty`?

Reply:
```
The type 'String' does not define the field, constructor or member 'IsEmpty'
```

Do I need to open anything or bribe someone to get it?

Reply:
`System.String.IsNullOrEmpty`. This is a bit ridiculous though. We should add an extension method to `String` at some point.